subprojects/logging/src/main/java/org/gradle/internal/featurelifecycle/LoggingDeprecatedFeatureHandler.java

```diff
@@ -43,39 +43,47 @@
 
     private final Set<String> messages = new HashSet<String>();
     private UsageLocationReporter locationReporter;
+
     private WarningMode warningMode;
+    private DeprecatedUsageBuildOperationProgressBroadaster buildOperationProgressBroadaster;
 
     public LoggingDeprecatedFeatureHandler() {
-        this(DoNothingReporter.INSTANCE);
-    }
-
-    public LoggingDeprecatedFeatureHandler(UsageLocationReporter locationReporter) {
-        this.locationReporter = locationReporter;
+        this.locationReporter = DoNothingReporter.INSTANCE;
     }
 
-    public void init(UsageLocationReporter reporter, WarningMode warningMode) {
+    public void init(UsageLocationReporter reporter, WarningMode warningMode, DeprecatedUsageBuildOperationProgressBroadaster buildOperationProgressBroadaster) {
         this.locationReporter = reporter;
         this.warningMode = warningMode;
+        this.buildOperationProgressBroadaster = buildOperationProgressBroadaster;
     }
 
     @Override
     public void featureUsed(FeatureUsage usage) {
-        if (messages.add(usage.getMessage())) {
+        String featureMessage = usage.formattedMessage();
+        if (messages.add(featureMessage)) {
             usage = usage.withStackTrace();
             StringBuilder message = new StringBuilder();
             locationReporter.reportLocation(usage, message);
             if (message.length() > 0) {
                 message.append(SystemProperties.getInstance().getLineSeparator());
             }
-            message.append(usage.getMessage());
-            logTraceIfNecessary(usage.getStack(), message);
+            message.append(featureMessage);
+            appendLogTraceIfNecessary(usage.getStack(), message);
             if (warningMode == WarningMode.All) {
                 LOGGER.warn(message.toString());
             }
         }
+        fireDeprecatedUsageBuildOperationProgress(usage);
+    }
+
+    private void fireDeprecatedUsageBuildOperationProgress(FeatureUsage usage) {
+        if (buildOperationProgressBroadaster != null) {
+            buildOperationProgressBroadaster.progress(usage, usage.withStackTrace().getStack());
```

Comment:
Can we make the stacktrace evaluation lazy? In my understanding, calculating stack trace is a relatively slow operation. My local experiment shows that invoking `FeatureUsage.withStacktrace().getStack()` can be as slow as 0.2ms with a 300-length stacktraces.

Reply:
We can if we use `Exception` under the hood to get the trace, because it has some internal mechanics for delaying generation of the trace.

@breskeby we should try this.